### ISSUE TITLE
Next.js cart storage 

### ISSUE THREAD

**Issue by peckpeck20** [notrab/react-use-cart#88]
Hi!

I'm using the library with Next.js. 

When trying to using the library in pages which rendered server-side, the content of the server doesn't match the client. It seems this could be related to the cart being stored using localStorage and this is not available on the server, only in the client. 

One thought it to use the storage prop, have you ran into this issue before ?

I tried using the storage prop and the following snipped fails with the following error. Could you provide an example of how the prop should be used and its initial value

```
function MyApp({ Component, pageProps }) {
  const [cart, setCart] = useState();
  return (
    <div>
      <CartProvider storage={() => [cart, setCart]}>
        <Layout>
          <Component {...pageProps} />;
        </Layout>
      </CartProvider>
    </div>
  );
}
```

![image](https://user-images.githubusercontent.com/22203676/129704150-91ecf587-b8e5-44f2-886b-936d506b52ec.png)

**Comment by notrab** [comment]
I've seen this before, but I haven't figured out a solution yet.

**Comment by debatz** [comment]
Very interested in this!

**Comment by notrab** [comment]
@debatz I was thinking we could do something like adding:

```js
useEffect(() => setMounted(true), []);
```

Inside of the `CartProvider`. It's a cheap way of checking if all is mounted, and if so, we can continue.

**Comment by debatz** [comment]
I haven't used this library yet, but was looking at it, and had found the same error using Uppy with NextJS.

My solution was to wrap the Uppy component with this:

```
ClientOnly.tsx

import dynamic from "next/dynamic"
import React from "react"

const ClientOnly = props => (
  <React.Fragment>{props.children}</React.Fragment>
)

export default dynamic(() => Promise.resolve(ClientOnly), {
  ssr: false
})
```

Your approach seems cleaner, adding just for context @notrab

**Comment by mattdjenkinson** [comment]
Hey, has there been a solution for this yet in NextJS? I'm getting a server and client mismatch on page reload.

**Comment by joshuabaker** [comment]
@notrab That approach is working for us.

```jsx
import { Fragment, useEffect, useState } from "react";
import { CartProvider as BaseCartProvider } from "react-use-cart";

export default function CartProvider({ children, ...props }) {
  const [mounted, setMounted] = useState(false);

  useEffect(() => setMounted(true));

  if (mounted) {
    return <BaseCartProvider {...props}>{children}</BaseCartProvider>;
  }

  return <Fragment>{children}</Fragment>;
}
```

**Comment by smarquez1** [comment]
I;m trying to use @joshuabaker 's solution but getting  an error calling setCartMetadata inside an `useEffect` hook. Any ideas on how to fix this?
```Unhandled Runtime Error
TypeError: setCartMetadata is not a function```

**Comment by notrab** [comment]
@smarquez1 do you have the provider set higher than where you're using that hook?

**Comment by smarquez1** [comment]
@notrab yes, I uploaded a [project](https://github.com/smarquez1/cart-test) to github with a minimal version of my main project. Here are the important parts:

https://github.com/smarquez1/cart-test/blob/main/components/CartProvider.tsx
https://github.com/smarquez1/cart-test/blob/main/pages/_app.tsx
https://github.com/smarquez1/cart-test/blob/main/pages/index.tsx

**Comment by joshuabaker** [comment]
This is happening because the `initialState` doesn’t include placeholder methods for the various setters.

https://github.com/notrab/react-use-cart/blob/f3f5d0216769ed54cdd1d98ec4ead2f6529dd30d/src/index.tsx#L55-L62

Until we mount we only have access to the values defined here, which don’t include noops.

Here’s another hacky workaround…

```jsx
const noop = () => {};
const { updateCartMetadata = noop } = useCart();
```

@notrab If you made use of [this open-source hook](https://github.com/streamich/react-use/blob/e53ca94a0b1f20270b0f75dc2ca1fecf1e119dde/src/useLocalStorage.ts) you might save the hassle on the mounting stuff.

**Comment by mdj-uk** [comment]
I used the solution above (https://github.com/notrab/react-use-cart/issues/88#issuecomment-1086117615) for a while, i.e. conditionally rendering `CartProvider` only once the app is mounted.

However that leads to a couple of issues 
1. it causes a re-render of all all components below `MyCartProvider` after react hydrates the page - i.e. a re-render of just about the whole app
2. sometimes trying to use `inCart()`, `isEmpty` etc. elsewhere in the app will fail, as `CartProvider` is not in place

So I think it's better to conditionally render the components/data which depend on the state of the cart instead.

Contrived example based on [new react docs](https://react.dev/reference/react-dom/client/hydrateRoot#handling-different-client-and-server-content):
```jsx
// _app.jsx
import { CartProvider } from 'react-use-cart'

function Application({ Component, pageProps }) {
  return (
    <CartProvider>
        <Component {...pageProps} />
    </CartProvider>
  )
}

export default Application;

// MyComponent.jsx
import { useCart } from "react-use-cart";
import { useEffect, useState } from "react";

export default function MyComponent() {
  const { totalUniqueItems } = useCart();
  const [isClient, setIsClient] = useState(false);

  useEffect(() => {
    setIsClient(true);
  }, []);

  if (isClient){
    return (
      <p>You have {totalUniqueItems} items in the cart</p>
    )
  }
  else {
    return (
      // stuff which will only render on the server
    )
  }
}
```

**Comment by monarcode** [comment]
i have this same problem, found another package instead that works with nextjs

https://github.com/mcnaveen/cart

**Comment by notrab** [comment]
If someone would like to help merge into support for App Router and other React contexts, please let me know. I don't have so much time these days to work on this as I've no sponsorship for this package currently.

**Comment by dani-z** [comment]
@notrab I am using the package inside App Router and seems to be working with no problems. Do you know of any issues?

**Comment by iamvinny** [comment]
> https://github.com/mcnaveen/cart

That package lacks a lot of features and haven't been updated in a while.

**Comment by garkuwa** [comment]
I've experienced problems in NextJs and just switched to Zustand https://github.com/pmndrs/zustand.
You can write your own cart hook in 50 lines of code with Zustand.  It also persists the data for you out of the box.
